support Symfony 4

ElectricMaxxx opened the pull request:
<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT, MAKE SURE YOU READ AND EDIT IT! -->

<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 2.x is for everything backwards compatible, like patches, features and deprecation notices
    - master is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/blob/2.x/CONTRIBUTING.md#the-base-branch
-->
I am targeting this branch, because i want to work on a new 2.x version

<!--
    Specify which issues will be fixed/closed.
    Remove it if this is not related.
-->

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: http://keepachangelog.com/
-->

<!-- REMOVE EMPTY SECTIONS -->
```markdown
### Added
- Symfony 4 support
```

## Subject

<!-- Describe your Pull Request content here -->
I want to allow Symfony 4.0 before making it required in travis.yaml, just to installi it in a testing project and to see what happens. I would add outcome as issues.

greg0ire commented:
> just to installi it in a testing project and to see what happens

I believe you can do that with `composer require symfony 4.0.3 as 3.4.0

jordisala1991 commented on composer.json:
```diff
@@ -49,8 +49,8 @@
-        "symfony/routing": "^2.8 || ^3.2",
-        "symfony/security-acl": "^2.8 || ^3.0"
+        "symfony/routing": "^2.8 || ^3.3 || ^4.0",
+        "symfony/security-acl": "^2.8 || ^3.3 || ^4.0"
```
security-acl only exist as 2.8 or 3.0

dbu commented:
now that you have a branch, you can add an entry to your composer.json to require that branch instead of the official repo of the phpcr admin bundle, then it will install with symfony 4. see https://getcomposer.org/doc/04-schema.md#repositories

i am -1 on merging this as is, as then people will install this with their symfony 4 and be surprised if it does not work at all. i only want to merge such a change once we also successfully test with sf 4.

dbu commented:
something like

```
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/ElectricMaxxx/allow_Symfony_4"
        },
...
     "require": 
    {
          "sonata-project/doctrine-phpcr-admin-bundle": "dev-allow_Symfony_4 as 2.1",
    } 
```

ElectricMaxxx commented:
@dbu i would say yes if we would release it, but we are on master which should be bleeding edge. I know that lots of people out there will try to use it, but exactly those people will report the issues. You remember the [content-bundle issue](https://github.com/symfony-cmf/content-bundle/issues/185) yesterday? Found through testing it on SF4 even if it is hard to bring it together. (got the same on resource-bundle now, and there was no deprecation notice)

ElectricMaxxx commented:
And btw: all other sonata bundle do have the `4.0` allowed, but sill no compatebility.

greg0ire commented:
Yeah but at least the tests pass

jordisala1991 commented:
You have to look at the allowed failures section. If the symfont one passes (and with a green build),  then the PR is ready.

ElectricMaxxx commented:
I also have make the Controller public.

jordisala1991 commented on composer.json:
```diff
@@ -49,7 +49,7 @@
         "symfony-cmf/resource-bundle": "^1.0",
         "symfony-cmf/testing": "^2.0.1",
```
You probably don't want to change this version, but you will need to release 2.1 to have some change to pass the dev-master build

ElectricMaxxx replied:
correct :-(
and there the chain begins: 
in testing component i wait for a release of phpcr-odm 2.0 to not require a dev dependency in my component. (@dbu  any info on the phpcr release lane?)

jordisala1991 replied:
I tried to make testing component compatible with sf 4 on the 2.0 branch, but as I saw it was already on 2.1 I abandoned that PR.

ElectricMaxxx replied:
Indeed, i also did some changes/improvements on testing (no breaks)

ElectricMaxxx replied:
but, we can say: lets release this one, when `testing` is released and so whem `phpcr-odm` is released?

dbu commented:
yeah, phpcr-odm... i think i will focus on that this week, have to get things moving. hard to say how much effort is needed (its a major version bump, so i want to be sure not to miss bc breaks that we have to do). the symfony changes will be small, its just console and maybe a few other such things. ideally we would probably have done those in the 1.x branch so that it would just be a minor release.

about merging this as is: what about https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/pull/521#issuecomment-360045823 - would that not work for now? i would prefer to use this branch over merging something that is completely broken, even if we don't release yet. otherwise we might as well add tests with symfony 4 and have a failing build on master...

ElectricMaxxx commented:
here is still some work for me, again.

greg0ire commented:
The build is broken

covex-nn commented:
I fixed `Bundle set sonata_admin_phpcr has not been registered` error in #526

covex-nn commented on composer.json:
```diff
@@ -47,9 +47,9 @@
         "symfony-cmf/resource": "^1.0",
         "symfony-cmf/resource-bundle": "^1.0",
-        "symfony-cmf/testing": "^2.0.1",
+        "symfony-cmf/testing": "^2.1",
```
i guess, you should not update this dependency: let `2.0.*` be used with PHP5.6 and `2.1.*` - with PHP7.1 and higher

ElectricMaxxx replied:
mhh, ok, ill try it.

ElectricMaxxx commented:
@covex-nn i removed the bundle set at all as we did not need it anymore in cmf. I also changed the behavior to fetch the kernel class in bootstrap process not to use `KERNEL_DIR`anymore, which is deprecated. So we have to use testing in version 2.1 and we have to set KERNEL_CLASS as env var then in `.travis.yml` I try to get this one working again locally

ElectricMaxxx commented on .travis.yml:
```diff
@@ -10,8 +10,6 @@ branches:
 
 php:
-  - '5.6'
-  - '7.0'
```
the changes i made here should be done in  `dev-kit`

covex-nn commented:
@ElectricMaxxx localy in your branch i added the following to `tests/Fixtures/App/config/sonata_phpcr_admin.yml` and all tests were passed:
```yaml
sonata_core:
    form:
        mapping:
            enabled: false
```
Othewise in `tests/Fixtures/App/var/logs/phpcr.log` there were errors like this:
> [2018-06-22 07:08:33] request.CRITICAL: ... "The "sonata.core.form.type.collection" service or alias has been removed or inlined when the container was compiled" ...

ElectricMaxxx commented:
@covex-nn ah good point, thanks. I will try to get it green at all.

ElectricMaxxx commented:
mhh, just the `proc_open()` issue is there, i have no clue where it comes from, maybe run the tests in a more verbose way

covex-nn commented:
It is from https://github.com/symfony-cmf/testing/issues/180 =)

ElectricMaxxx commented:
Ohhh my good it is green on travis ... yeeaaaaa 
i would need a hand of @greg0ire  or an other Contributer to move changes in `Makefile` and `.travis.yml` into dev-kit update.

covex-nn commented on .travis.yml:
```diff
@@ -10,14 +10,11 @@ branches:
 
 php:
-  - '5.6'
-  - '7.0'
```
@ElectricMaxxx i found that you droped support for PHP5.6 here https://github.com/symfony-cmf/testing/pull/146 and for PHP7.0 here https://github.com/symfony-cmf/testing/pull/155 in `symfony-cmf/testing`. But this bundle still supports that versions!

So, i created a PR to test how `symfony-cmf/testing` works with PHP 5.6 and PHP 7.0 https://github.com/symfony-cmf/testing/pull/181. I solved a "composer and out or memory" issue, and travis is green these. See for yourself https://travis-ci.org/symfony-cmf/testing/builds/396024428

covex-nn commented on tests/Fixtures/App/Admin/ContentAdmin.php:
```diff
@@ -124,7 +124,7 @@ protected function configureFormFields(FormMapper $formMapper)
                 'Sonata\AdminBundle\Form\Type\ModelType',
```
Please, use FQCN here

covex-nn commented on tests/Fixtures/App/Admin/ContentAdmin.php:
```diff
@@ -124,7 +124,7 @@ protected function configureFormFields(FormMapper $formMapper)
                 [
                     'property' => 'title',
-                    'class' => 'Sonata\DoctrinePHPCRAdminBundle\Tests\Resources\Document\Content',
+                    'class' => 'Sonata\DoctrinePHPCRAdminBundle\Tests\Fixtures\App\Document\Content',
```
Please, use FQCN here

covex-nn commented on tests/WebTest/Admin/ContentAdminTest.php:
```diff
@@ -17,13 +17,13 @@ class ContentAdminTest extends BaseTestCase
     public function setUp()
     {
-        $this->db('PHPCR')->loadFixtures(['Sonata\DoctrinePHPCRAdminBundle\Tests\Resources\DataFixtures\Phpcr\LoadTreeData']);
+        $this->db('PHPCR')->loadFixtures(['Sonata\DoctrinePHPCRAdminBundle\Tests\Fixtures\App\DataFixtures\Phpcr\LoadTreeData']);
```
Please, use FQCN here

covex-nn commented on tests/WebTest/Dashboard/TreeBrowserTest.php:
```diff
@@ -20,7 +20,7 @@ class TreeBrowserTest extends BaseTestCase
     public function setUp()
     {
-        $this->db('PHPCR')->loadFixtures(['Sonata\DoctrinePHPCRAdminBundle\Tests\Resources\DataFixtures\Phpcr\LoadTreeData']);
+        $this->db('PHPCR')->loadFixtures(['Sonata\DoctrinePHPCRAdminBundle\Tests\Fixtures\App\DataFixtures\Phpcr\LoadTreeData']);
```
Please, use FQCN here

covex-nn commented:
@ElectricMaxxx i propose the following plan:

1. We stabilize `2.x` branch with #527 
2. You make `symfony-cmf/testing` compatible with php >= 5.6. It is possible, see https://github.com/symfony-cmf/testing/pull/181
3. We update `symfony-cmf/testing` to run `doctrine:phpcr:init:dbal` and `doctrine:phpcr:repository:init` from `.travis/before_script_test.sh`, see https://github.com/symfony-cmf/testing/pull/182
4. You release `symfony-cmf/testing` v2.1.2, update constraint in composer.json in this PR, revert changes in Makefile and put `vendor/symfony-cmf/testing/bin/console` commands to `.travis/before_script_test.sh`. From this point travis should be happy =)
5. We update `sonata-project/dev-kit` - 2 console commands will be added to `.travis/before_script_test.sh`
6. This PR will be squashed, rebased and tested again with new dev-kit changes, then accepted and merged
7. :tada: 

What do you think?

ElectricMaxxx commented:
Sounds awesome. I did some tiny comments in your PR

covex-nn commented on phpunit.xml.dist:
```diff
@@ -33,20 +25,16 @@
-
     <php>
         <ini name="precision" value="8"/>
+        <env name="KERNEL_CLASS" value="Sonata\DoctrinePHPCRAdminBundle\Tests\Fixtures\App\Kernel" />
```
Please, add `<server name="KERNEL_DIR" value="tests/Fixtures/App"/>` too

covex-nn replied:
With `2.1.2` release KERNEL_DIR was required, but https://github.com/symfony-cmf/testing/pull/183 made possible to set `KERNEL_CLASS` only )

covex-nn commented:
Somehow `.travis/before_script_test.sh` is not executable =( and that is why `phpcr` was not prepared with `if [ -x .travis/before_script_${TARGET}.sh ]; then .travis/before_script_${TARGET}.sh; fi;`

greg0ire commented on .travis/before_script_test.sh:
```diff
@@ -1,3 +1,9 @@
+
+@echo '+++ create PHPCR +++'
+${CONSOLE} doctrine:phpcr:init:dbal --drop --force
+${CONSOLE} doctrine:phpcr:repository:init
```
Please configure your IDE to always add a newline at end of file, this will make the red pictogram disappear.

dbu commented:
i think the non installable places are because php < 7.1. lets adjust the matrix to only build for php 7.1 and 7.2, and also change composer.json to require at least 7.1. ok?

ElectricMaxxx commented:
@dbu yea lets bring it to its end

covex-nn commented:
@dbu @ElectricMaxxx please, release `symfony-cmf/testing` with php>=5.6 support, it does not support older versions only in it's composer.json, see https://github.com/symfony-cmf/testing/pull/181 and https://github.com/symfony-cmf/dev-kit/pull/4

After this there will be only few simple steps to release new version of `SonataDoctrinePhpcrAdminBundle`, see my comment https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/pull/521#issuecomment-401170553

Anyway, dropping support for php<7.1 is an option too. But is it correct to drop it without new major release?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

ElectricMaxxx commented:
Yes i will do tonight.

ElectricMaxxx commented:
It went green, but i put the class into a constant defined in travis.yml, when i do same by doing a simple export in i.e. `before_script.sh`, it breaks. The reason is: the document mapping paths are defined while bootstrapping, when the kernel is resolved. So if the kernel is not resolved through the class name, the mapping isn't created. What to do now?

ElectricMaxxx commented:
Ohhh yes it works in `before_script.sh` so then lets have a party 
@dbu @covex-nn @kunicmarko20 somebody would like to review?

ElectricMaxxx commented on Makefile:
```diff
@@ -2,6 +2,8 @@
 #
 # It's auto-generated by sonata-project/dev-kit package.
 
+export KERNEL_CLASS=Sonata\DoctrinePHPCRAdminBundle\Tests\Fixtures\App\Kernel
```
ahhh is still have it here, i hope it will work without it.

ElectricMaxxx commented on composer.json:
```diff
@@ -23,19 +23,20 @@
+        "symfony/http-kernel": "^2.8 || ^3.2 || ^4.0",
+        "symfony/options-resolver": "^2.8 || ^3.2 || ^4.0",
+        "symfony/property-access": "^2.8 || ^3.2 || ^4.0",
+        "symfony/symfony": "^3.4"
```
this line needs to go.

ElectricMaxxx commented:
BTW: it was not my idea to have those style changes. The linter forces me to do so.

ElectricMaxxx commented:
@dbu  i did the change on 2.x https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/pull/528

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

ElectricMaxxx commented:
Interessting, it breaks with that value:
![selection_353](https://user-images.githubusercontent.com/2905834/44890132-50e70c00-acd9-11e8-9f7e-bf2b4c7e25af.jpg)

ElectricMaxxx commented:
@dbu you ever seen that issue? Seems as the class for the definition of doctrine namespaces (the cached version) is wrong, istn't it?

dbu commented on docs/reference/document_tree.rst:
```diff
@@ -220,7 +220,7 @@ like this:
 
         {% render(controller(
-            'sonata.admin.doctrine_phpcr.tree_controller:treeAction',
+            'sonata.admin.doctrine_phpcr.tree_controller::treeAction',
```
afaik this only works with newer symfony versions. maybe add a note after that block that on symfony < 3.4 people should use single `:`?

dbu commented on src/Form/Type/ChoiceFieldMaskType.php:
```diff
@@ -62,7 +63,7 @@ public function configureOptions(OptionsResolver $resolver)
     public function getParent()
     {
-        return 'choice';
+        return ChoiceType::class;
```
afaik this does not work with symfony 2.8, or does it? if it does not, lets raise the minimal symfony version to 3.4 to keep things manageable.

dbu replied:
are you sure that this works with symfony 2?

dbu commented on src/Resources/config/routing/tree.xml:
```diff
@@ -1,7 +1,7 @@
 <routes xmlns="http://symfony.com/schema/routing" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/routing http://symfony.com/schema/routing/routing-1.0.xsd">
     <route id="sonata_admin_tree" path="/tree">
-        <default key="_controller">sonata.admin.doctrine_phpcr.tree_controller:treeAction</default>
+        <default key="_controller">sonata.admin.doctrine_phpcr.tree_controller::treeAction</default>
```
i think this is another reason to raise the min symfony version to 3.4. otherwise we need a parameter for this or 2 different files to also have the single colon syntax. and thats not worth it imho.

dbu replied:
does this really work with symfony 2?

dbu commented:
i think the symfony 2.8 build working probably means we have a low test coverage. or i am wrong about since when the `::` for controller as services is supported.

i would bump php to ^7.1 as well. its going to be a new version, no point supporting old versions if they create problems. (its not "simply" anymore like you wrote in the merge request title, of course. almost nothing ever is :-/ )

ElectricMaxxx commented:
Except that fintci thing it goes green now. The php version change is done through dev-kit also: https://github.com/sonata-project/dev-kit/pull/427

greg0ire commented:
@ElectricMaxxx please apply the FlintCI patch, the merge button is disabled otherwise (yes, we are that strict about it :P )

greg0ire commented:
`wget https://flintci.io/repositories/75/analyses/11470/diff && patch -p1 < diff && rm diff`

greg0ire commented:
Finally! A green build! Good job!

rande commented:
Just ping me on slack when the PR is ready, I will merge it ;)

Thanks for the work and patience.

OskarStark commented:
@ElectricMaxxx could you please finish this PR?

ElectricMaxxx commented:
Ohh yes, lets do it.

ElectricMaxxx commented on tests/Fixtures/App/Admin/ContentAdmin.php:
```diff
@@ -18,7 +18,12 @@
+use Sonata\CoreBundle\Form\Type\CollectionType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Sonata\AdminBundle\Form\Type\ModelType;
+use Symfony\Cmf\Bundle\TreeBrowserBundle\Form\Type\TreeSelectType;
```
would be nice, when code style is the only issue :-)

ElectricMaxxx commented on docs/reference/document_tree.rst:
```diff
@@ -231,14 +231,18 @@ like this:
         )) ?>
 
+.. note::
+    To use the configuration for Symfony < 3.4 you you should use the single colon (:) notation to define controller
```
good?

ElectricMaxxx commented:
i am confused now: i got an exception that i user `choices_as_values` as an option in my type, but i don't. I there a chance i inheret thtat options?

dbu commented on docs/reference/document_tree.rst:
```diff
@@ -231,14 +231,20 @@ like this:
+.. note::
+    To use the configuration for Symfony < 3.4 you should use the single colon (:) notation to define controller
+    actions: ``sonata.admin.doctrine_phpcr.tree_controller:treeAction``. Have a look at the `Symfony documentation`_
+    (here at the example for version 3.1).
```
i would leave away the link (so the last sentence). and s/you should use/have to use/. with double colon it would simply be broken, so its not just "should"

OskarStark commented on .travis/before_install_test.sh:
```diff
@@ -5,8 +5,6 @@ PHP_INI_DIR="$HOME/.phpenv/versions/$(phpenv version-name)/etc/conf.d/"
 echo "memory_limit=3072M" >> "$TRAVIS_INI_FILE"
 
-
-
```
Can you please remove this diff?

OskarStark commented on .travis/before_script_test.sh:
```diff
@@ -1,3 +1,9 @@
 #!/usr/bin/env bash
 set -ev
+TESTING_SCRIPTS_DIR=vendor/symfony-cmf/testing/bin
```
IMO this file would be overridden by dev kit afterwards...

And can you switch back the chmod of this file please?

ElectricMaxxx replied:
No, i can not. Have the files non executeable, which is 644, breaks the check and validation for their existence. So a `+x` solved it:

```bash
maximilian@maximilian-ThinkPad-T470p:~/OpenSource/SonataProject/SonataDoctrinePhpcrAdminBundle$ sudo chmod +x .travis/*
maximilian@maximilian-ThinkPad-T470p:~/OpenSource/SonataProject/SonataDoctrinePhpcrAdminBundle$ if [ -x .travis/before_install_test.sh ]; then echo "GEHT"; fi;
GEHT
```

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

ElectricMaxxx commented:
uuuh and not it si geen again. 
And i won't move any char anymore :-)
So pls. merge and move open changes into a separate Issue pls.

dbu commented on composer.json:
```diff
@@ -21,21 +21,26 @@
+    },
+    "conflict": {
+        "sebastian/environment": "<1.3.4",
+        "sebastian/exporter": "<2.0.0"
```
these are not real conflicts i guess, but constraints to make the lowest verison tests not fail. if so, they should go into require-dev as this would prevent installing this bundle in a project with a lower version of those sebastian packages.

OskarStark replied:
We should move them to `require-dev` 👍

dbu commented:
see #535 for the change on composer.json

dbu commented on docs/reference/document_tree.rst:
```diff
@@ -220,7 +220,7 @@ like this:
 
         {% render(controller(
-            'sonata.admin.doctrine_phpcr.tree_controller:treeAction',
+            'sonata.admin.doctrine_phpcr.tree_controller::tree',
```
don't we still need the "Action" suffix with the double colon notation?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

ElectricMaxxx commented:
OHH nice to face that issue hier directly ... :-) @greg0ire @kunicmarko20  i faced that issue (failing tests) today also, when i udated an application. 
It seems as core-bundle points to Runtime, which isn't there ( i saw that you create a twit-extension package) in my app i had to require your new package dev wise and register the runtime with a tag.

kunicmarko20 commented:
I think "simply" could be removed from the PR name? :sweat_smile:

OskarStark commented on .travis/install_test.sh:
```diff
@@ -7,7 +7,7 @@ mkdir --parents "${HOME}/bin"
     PHPUNIT_VERSION=7
 else
-    PHPUNIT_VERSION=5.7
+    PHPUNIT_VERSION=5.7.26
```
this will be overwritten by dev-kit, please add it to the devkit PR

ElectricMaxxx replied:
Isn't it?

OskarStark replied:
Indeed, I saw the PR afterwards

OskarStark commented on UPGRADE-3.0.md:
```diff
@@ -0,0 +1,2 @@
+UPGRADE FROM 2.x to 3.0
```
do we need this file?

ElectricMaxxx replied:
Isn't it?

ElectricMaxxx commented:
It is, I hope. -------- Originalnachricht --------Betreff: Re: [sonata-project/SonataDoctrinePhpcrAdminBundle] simply allow SF4 installation (#521)Von: Oskar Stark An: sonata-project/SonataDoctrinePhpcrAdminBundle Cc: Maximilian Berghoff ,Mention @OskarStark commented on this pull request.



In .travis/install_test.sh:
> @@ -7,7 +7,7 @@ mkdir --parents "${HOME}/bin"
 if [ "${TRAVIS_BRANCH}" = 'master' ]; then
     PHPUNIT_VERSION=7
 else
-    PHPUNIT_VERSION=5.7
+    PHPUNIT_VERSION=5.7.26

this will be overwritten by dev-kit, please add it to the devkit PR

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/sonata-project/SonataDoctrinePhpcrAdminBundle","title":"sonata-project/SonataDoctrinePhpcrAdminBundle","subtitle":"GitHub repository","main_image_url":"https://github.githubassets.com/images/email/message_cards/header.png","avatar_image_url":"https://github.githubassets.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle"}},"updates":{"snippets":[{"icon":"PERSON","message":"@OskarStark commented on #521"}],"action":{"name":"View Pull Request","url":"https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/pull/521#pullrequestreview-185105694"}}}
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/pull/521#pullrequestreview-185105694",
"url": "https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/pull/521#pullrequestreview-185105694",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

ElectricMaxxx commented:
@dbu you had some time to get my last changes into your branche?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

OskarStark commented on .gitignore:
```diff
@@ -5,3 +5,5 @@
 phpunit.xml
 tests/Resources/app/var
+tests/Fixtures/App/var/logs/
+tests/Fixtures/App/var/cache/
```
please add newline

OskarStark commented on .travis/check_relevant_docs.sh:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+#!/usr/bin/env sh
```
If we would like to keep this, this change should be done against dev-kit

ElectricMaxxx replied:
no was my fault while mergin

OskarStark commented on before_script_test.local.sh:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
```
dan, this must be in root folder... we should change this to .travis folder

cc @greg0ire

ElectricMaxxx replied:
yea i struggled over it :-)

OskarStark commented on tests/Fixtures/App/var/cache/AppPhpcrDebugProjectContainer.xml.meta:
```diff
@@ -0,0 +1 @@
+a:108:{i:0;C:50:"Symfony\Component\Config\Resource\ComposerResource":106:{a:1:{s:79:"/home/maximilian/OpenSource/SonataProject/SonataDoctrinePhpcrAdminBundle/vendor";i:1541711911;}}i:1;C:46:"Symfony\Component\Config\Resource\FileResource":111:{s:102:"/home/maximilian/OpenSource/SonataProject/SonataDoctrinePhpcrAdminBundle/tests/Fixtures/App/Kernel.php";}i:2;C:46:"Symfony\Component\Config\Resource ...
```
Please add new line

ElectricMaxxx replied:
This one should be ignored instead. I mean the complete file.

OskarStark commented on tests/Fixtures/App/var/cache/AppPhpcrDebugProjectContainerDeprecations.log:
```diff
@@ -0,0 +1 @@
+a:0:{}
```
Please add new line

OskarStark commented on tests/Fixtures/App/var/cache/annotations.map:
```diff
@@ -0,0 +1,4 @@
+<?php return array (
+  0 => 'Symfony\\Bundle\\FrameworkBundle\\Controller\\AbstractController',
+  1 => 'Symfony\\Bundle\\FrameworkBundle\\Controller\\Controller',
+);
```
Please add new line

ElectricMaxxx commented on .travis.yml:
```diff
@@ -31,7 +31,7 @@ cache:
   global:
     - PATH="$HOME/.local/bin:$PATH"
-    - SYMFONY_DEPRECATIONS_HELPER=weak
+    - SYMFONY_DEPRECATIONS_HELPER=weak_vendors
```
this one is required? or shouldn't that excluder other deprecations?

OskarStark replied:
Ping @greg0ire

greg0ire replied:
Revert to `weak`, there is a bug with `weak_vendors` apparently.

greg0ire commented:
There is an issue with the `weak_vendors` mode: it wrongly detects some of the deprecations as coming from vendors. Please revert to `weak` mode and the build should finally be green. @core23, please advise on how to fix those deprecations, they are about services you deprecated, but we don't know in favor of what.

greg0ire commented on composer.json:
```diff
@@ -23,19 +23,22 @@
+        "symfony/options-resolver": "^2.8 || ^3.2 || ^4.0",
+        "symfony/property-access": "^2.8 || ^3.2 || ^4.0",
+        "sebastian/environment": "^1.3.4",
+        "sebastian/exporter": "^2.0"
```
This is not sorted hence the FlintCI issue. Also, why did you touch this? Also, should really this be in the `require` section, as opposed to the `require-dev` section?

ElectricMaxxx commented:
@greg0ire unfortunatel that hint does not fix it :-(

greg0ire commented:
Great! Next thing is to use `^2.8.8 || ^3.2 || ^4.0` for `symfony/form`!

ElectricMaxxx commented:
ohh now i understand your game :-)

ElectricMaxxx commented:
if mine is green I won't wait for a second to push the merge button. You can rebase on 2.x then an do you stuff afterwards also, or?

greg0ire commented:
Hold your breath, the build is starting!

greg0ire commented on composer.json:
```diff
@@ -49,9 +54,9 @@
-        "symfony-cmf/testing": "^2.1.11",
-        "symfony/phpunit-bridge": "^4.2",
-        "symfony/routing": "^2.8 || ^3.2",
+        "symfony-cmf/testing": "^2.1.8",
```
Why did you downgrade that one???

ElectricMaxxx replied:
sorry, this was old, before your changes or i fucked it up in my rebase

greg0ire commented on composer.json:
```diff
@@ -49,9 +54,9 @@
-        "symfony/phpunit-bridge": "^4.2",
-        "symfony/routing": "^2.8 || ^3.2",
+        "symfony-cmf/testing": "^2.1.8",
+        "symfony/phpunit-bridge": "^4.0",
```
Or that one?

greg0ire commented:
Still not good, but no errors, only failures now. Please revert the changes for cmf testing and the phpunit bridge

OskarStark commented:
Nearly one year ago, you started this PR 😄

greg0ire commented on src/Block/TreeBlockService.php:
```diff
@@ -45,7 +45,7 @@ public function __construct($name, EngineInterface $templating, array $defaults
      * NOOP as there is nothing to edit.
      */
-    public function buildEditForm(FormMapper $form, BlockInterface $block)
+    public function buildEditForm(FormMapper $form, BlockInterface $block): void
```
I don't think we do… maybe in the past we did?

ElectricMaxxx replied:
some of the style checker forced me to do so :-( some time ago.

ElectricMaxxx replied:
@greg0ire all or just in this file?

OskarStark replied:
in all files please

ElectricMaxxx replied:
ohh quite less changes now.

ElectricMaxxx replied:
Done!

OskarStark commented on docs/reference/document_tree.rst:
```diff
@@ -231,14 +231,19 @@ like this:
                 ),
         )) ?>
 
+.. note::
```
Please add a blank line after `.. note::` directive

OskarStark commented on docs/reference/document_tree.rst:
```diff
@@ -231,14 +231,19 @@ like this:
                     'selected' => $menuNodeId,
                     '_locale'  => $app->getRequest()->getLocale()
                 ),
         )) ?>
```
```suggestion
        ]);
```

OskarStark commented on docs/reference/document_tree.rst:
```diff
@@ -231,14 +231,19 @@ like this:
         <?php echo $view['actions']->render(new ControllerReference(
-                'sonata.admin.doctrine_phpcr.tree_controller:treeAction',
+                'sonata.admin.doctrine_phpcr.tree_controller::tree',
                 array(
```
```suggestion
                [
```

OskarStark commented on src/Block/TreeBlockService.php:
```diff
@@ -14,26 +14,21 @@
     protected $defaults;
 
-    /**
-     * @param string          $name
```
this line should be kept, until we can use a typehint `string` in the parameter signature

OskarStark commented on src/Builder/ListBuilder.php:
```diff
@@ -57,7 +58,7 @@ public function buildField($type, FieldDescriptionInterface $fieldDescription, A
         if (null == $type) {
             $guessType = $this->guesser->guessType($admin->getClass(), $fieldDescription->getName(), $admin->getModelManager());
-            $fieldDescription->setType($guessType->getType());
+            $fieldDescription->setType(($guessType instanceof TypeGuess ? $guessType->getType() : null));
```
are the additional brackets needed? 🤔 

Or does this work, too?:
```suggestion
            $fieldDescription->setType($guessType instanceof TypeGuess ? $guessType->getType() : null);
```

greg0ire replied:
It definitely should

ElectricMaxxx replied:
in free living conditions i would add them for better readability. Here it is an edge case i would say, it should work at the end, but i will fix it.
btw: you know a good boolean cast in javascript? `!!(condition)` *g* that is quite awesome writing angular expressions.

OskarStark commented on tests/Fixtures/App/Admin/ContentAdmin.php:
```diff
@@ -98,41 +103,43 @@ protected function configureFormFields(FormMapper $formMapper)
                 ['property' => 'title', 'multiple' => true, 'expanded' => false]
             )
-            ->add('parentDocument', 'Symfony\Cmf\Bundle\TreeBrowserBundle\Form\Type\TreeSelectType', [
+            ->add('parentDocument',
```
```suggestion
            ->add(
                'parentDocument',
```

OskarStark commented on tests/Fixtures/App/config/bundles.php:
```diff
@@ -0,0 +1,25 @@
+<?php
+
+/*
```
please use the Sonata project header here instead of the Symfony CMF one

greg0ire replied:
The right header can be found here:
https://github.com/sonata-project/SonataDoctrinePhpcrAdminBundle/blob/03b03959b7304c86582f3faf7ec8504b2e468d2b/.php_cs.dist#L10-L15

OskarStark commented on tests/Unit/Builder/ListBuilderTest.php:
```diff
@@ -58,14 +59,14 @@ class ListBuilderTest extends TestCase
     public function setUp()
     {
-        $this->guesser = $this->createMock('\Sonata\AdminBundle\Guesser\TypeGuesserInterface', [], []);
+        $this->guesser = $this->createMock(TypeGuesserInterface::class, [], []);
```
IIRC, there are no more parameters for method: `createMock()`:

```php
    /**
     * Returns a test double for the specified class.
     *
     * @param string|string[] $originalClassName
     *
     * @throws Exception
     * @throws \InvalidArgumentException
     */
    protected function createMock($originalClassName): MockObject
    {
        return $this->getMockBuilder($originalClassName)
                    ->disableOriginalConstructor()
                    ->disableOriginalClone()
                    ->disableArgumentCloning()
                    ->disallowMockingUnknownTypes()
                    ->getMock();
    }
```

same for the following `createMock()`-calls

greg0ire replied:
```suggestion
        $this->guesser = $this->createMock(TypeGuesserInterface::class);
```

OskarStark commented on tests/Unit/Datagrid/PagerTest.php:
```diff
@@ -16,14 +16,15 @@
         $this->pager = new Pager(10);
 
-        $this->proxyQuery = $this->getMockBuilder('Sonata\DoctrinePHPCRAdminBundle\Datagrid\ProxyQuery')
+        $this->proxyQuery = $this->getMockBuilder(ProxyQuery::class)
```
please use `createMock()` here instead

greg0ire replied:
@ElectricMaxxx did not introduce `getMockBuilder`, maybe this could be done in a separate PR?

OskarStark replied:
It could definitely, but he is modifiying this line anyway. No KO criteria, I would do it afterwards, too if he don't want to do it 👍

OskarStark commented on tests/Unit/Datagrid/ProxyQueryTest.php:
```diff
@@ -31,10 +33,10 @@ class ProxyQueryTest extends TestCase
     public function setUp()
     {
-        $this->qb = $this->getMockBuilder('Doctrine\ODM\PHPCR\Query\Builder\QueryBuilder')
+        $this->qb = $this->getMockBuilder(QueryBuilder::class)
```
please use `createMock()` here

OskarStark commented on tests/Unit/Datagrid/ProxyQueryTest.php:
```diff
@@ -31,10 +33,10 @@ class ProxyQueryTest extends TestCase
             ->disableOriginalConstructor()
             ->getMock();
-        $this->query = $this->getMockBuilder('Doctrine\ODM\PHPCR\Query\Query')
+        $this->query = $this->getMockBuilder(Query::class)
```
please use `createMock()` here

OskarStark commented on tests/Unit/Datagrid/ProxyQueryTest.php:
```diff
@@ -115,7 +117,7 @@ public function testExecute()
     public function testGetAndSetDocumentManager()
     {
-        $dm = $this->getMockBuilder('Doctrine\\ODM\\PHPCR\\DocumentManager')
+        $dm = $this->getMockBuilder(DocumentManager::class)
```
please use `createMock()` here

OskarStark commented on tests/Unit/Filter/BaseTestCase.php:
```diff
@@ -40,7 +40,7 @@ public function setUp()
         $this->qbTester = new QueryBuilderTester($this->qb);
 
-        $this->proxyQuery = $this->getMockBuilder('Sonata\DoctrinePHPCRAdminBundle\Datagrid\ProxyQuery')
+        $this->proxyQuery = $this->getMockBuilder(ProxyQuery::class)
```
please use `createMock()` here

OskarStark commented on tests/Fixtures/App/Kernel.php:
```diff
@@ -0,0 +1,45 @@
+    {
+        $this->registerConfiguredBundles();
+        $this->requireBundleSet('default');
+        $this->requireBundleSets(['phpcr_odm',]);
```
Please remove the comma

OskarStark commented on docs/reference/document_tree.rst:
```diff
@@ -231,14 +231,20 @@ like this:
     .. code-block:: php
 
         <?php echo $view['actions']->render(new ControllerReference(
```
```suggestion
        echo $view['actions']->render(new ControllerReference(
```

OskarStark commented on docs/reference/document_tree.rst:
```diff
@@ -231,14 +231,20 @@ like this:
                     '_locale'  => $app->getRequest()->getLocale()
-                ),
+                ],
         )) ?>
```
```suggestion
        ));
```

OskarStark commented on docs/reference/document_tree.rst:
```diff
@@ -231,14 +231,20 @@ like this:
+                [
                     'root'     => $basePath . '/menu',
                     'selected' => $menuNodeId,
                     '_locale'  => $app->getRequest()->getLocale()
```
```suggestion
                    '_locale'  => $app->getRequest()->getLocale(),
```

OskarStark commented on docs/reference/filter_field_definition.rst:
```diff
@@ -202,8 +202,8 @@ To make the filter case insensitive, use the ``compare_case_insensitive`` option
-                ->add('author', 'doctrine_phpcr_string', array('compare_case_insensitive' => true))
+                ->add('title', 'doctrine_phpcr_string', ['compare_case_insensitive' => true])
+                ->add('author', 'doctrine_phpcr_string', ['compare_case_insensitive' => true])
                 ->add('label', 'doctrine_phpcr_string')
```
Statt `doctrine_phpcr_string` sollte ....::class genutzt werden inklusive dem use use Statement

ElectricMaxxx replied:
I changed them all, but this can be tricky again now.

OskarStark commented on docs/reference/form_field_definition.rst:
```diff
@@ -23,11 +23,8 @@ you can use the corresponding option in the form field definition:
-                        'placeholder' => 'No author selected'
-                    ))
+                    ->add('enabled', null, ['required' => false])
+                    ->add('author', 'sonata_type_model_list', [], ['placeholder' => 'No author selected'])
```
```suggestion
                    ->add('author', 'sonata_type_model_list', [], [
                        'placeholder' => 'No author selected',
                    ])
```

OskarStark commented on tests/Fixtures/App/config/bundles.php:
```diff
@@ -0,0 +1,23 @@
+<?php
+
+/*
+ * This file is part of the Sonata Project package.
```
```suggestion
 * This file is part of the Sonata Project package.
 
```

OskarStark commented on tests/Fixtures/App/config/bundles.php:
```diff
@@ -0,0 +1,23 @@
+
+/*
+ * This file is part of the Sonata Project package.
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
```
```suggestion
 * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
 
```

greg0ire commented:
Who canceled the build and why?

OskarStark commented:
If Travis is happy, lets merge it!

ElectricMaxxx commented:
*sing* 
> If travis is happy and you know it, merge your code ...

greg0ire commented:
Oh no

ElectricMaxxx commented:
*sing*
> haleluja, haleluja

greg0ire commented:
Congrats @ElectricMaxxx ! You did it!!!

dbu commented:
woohoo. never say "simple" :-D 
thanks max!

OskarStark commented:
Great work 🚀
